### PR TITLE
feat(graphql): add graphql-core integration

### LINKED ISSUE #925: Add graphql support
Are there plans to support graphql/graphene in the python implementation of ddtrace? It seems the JS library supports graphql and gives us insight into which graphql query/mutation is called with each request along with it's parameters.

### PR DESCRIPTION
**Pull request by mabdinur**
# Graphql-core Integration

## Overview 

This integration supports [graphql-core 3](https://github.com/graphql-python/graphql-core) and [graphql-core 2](https://github.com/graphql-python/graphql-core-legacy). By transitive dependencies executing graphql schemas in `graphene>2.0` is also supported.  This integration provides instrumentation for executing graphql queries using `graphql.graphql()`, `graphql.graphql_sync()`, `graphql.execute()`, and `graphql.execution.execute()`, `graphql.execution.executor.execute()`, `graphql.execute_sync()`, and `graphql.execution.execute_sync()`. 

Resolves #925. 

## Implementation Details 

[graphql-core](https://graphql-core-3.readthedocs.io/en/stable/modules/graphql.html#top-level-functions) this integration provides support for parsing, building, validating, and executing graphql schemas. It also provides additional insight into resolving [GraphQLFields](https://graphql.org/learn/queries/#fields).

The methods listed below are the main parts of the public api which are fully supported by the ddtrace library. All other methods listed in [graphql.__init__](https://github.com/graphql-python/graphql-core/blob/4a63f97bd8dc676a984bf81787a62cab7848d2ae/src/graphql/__init__.py) will not produce any instrumentation.

### graphql.graphql() and graphql.graphql_sync()

When [graphql.graphql()](https://github.com/graphql-python/graphql-core-legacy/blob/v2.0.0/graphql/graphql.py#L33) or [graphql.graphql_sync](https://github.com/graphql-python/graphql-core/blob/v3.0.0/src/graphql/graphql.py#L133)  (in `graphql-core>=3`) is called the following trace is produced:
- span name: graphql.query, resource: graphql source string
    - span name: graphql.parse, resource: graphql.parse
       converts the body of an incoming graphql request into a [graphql document](https://github.com/graphql-python/graphql-core/blob/main/src/graphql/graphql.py#L173)
    - span name: graphql.validate, resource: graphql.validate
       ensures a graphql schema is [syntactically correct and can be executed in a given context](https://github.com/graphql-python/graphql-core/blob/main/src/graphql/graphql.py#L173)
    - span name: graphql.execute, resource: graphql source string
       executes a valid graphql schema and returns a [response](https://github.com/graphql-python/graphql-core/blob/main/src/graphql/graphql.py#L173)
         - span name: graphql.resolve, resource: name of graphql field
         - span name: graphql.resolve, resource: name of graphql field
         -  ........ (a span is produced for every resolved field)
            Graphql queries can be complex and often resolve multiple Graphql fields in single request. Instrumenting resolvers provides insight on the execution of individual graphql fields and can highlight bottlenecks in fetching specific types of data. However for some applications this can produce a large number of spans. 

Note: If a GraphqlError is raised while parsing or validating a graphql query then `graphql.execute` will not be called and instrumented. For this reason we add the graphql source string to the graphql.query span AND the graphql.execute span.

### graphql.execute() and graphql.execution.execute()

When `execute()` is called the following trace is produced:
- span name: graphql.execute, resource: graphql source string
    - span name: graphql.resolve, resource: name of graphql field
    - span name: graphql.resolve, resource: name of graphql field
    -  ........ (a span is produced for every resolved field)
 
Note: `graphql-core>=3.1` introduced `execute_sync`. Using `graphql.execute_sync()` and `graphql.execution.execute_sync()` should produce the same trace as `graphql.execute()`.

## Future Work
 - add `graphql.source`, `graphql.operation.type` and `graphql.operation.name` span tags to maintain feature parity with the .NET tracer. 
 - instrument the execution of graphql [subscriptions](https://graphql.org/blog/subscriptions-in-graphql-and-relay/) in [graphql-core 3](https://github.com/graphql-python/graphql-core/blob/v3.2.1/src/graphql/execution/subscribe.py#L29). Due to how we patch `graphql.execution.executor.execute()` in [graphql-core 2](https://github.com/graphql-python/graphql-core-legacy/blob/02605b1adce7b287fa9ee6beacd735882954159a/graphql/execution/executor.py#L51) this version of the library generates graphql.execute and graphql.resolver spans when subscriptions are executed. For consistency we should extend this functionality to graphql-core 3 and offically support graphql subscriptions. 
    - sample patching:
       ```
        # patch execute function used in subscribe
        if graphql_version < (3,0):
            # graphql.execution.executor.execute() is used in graphql.execution.executor.subscribe()
            # this method is patched above.
            pass
        elif graphql_version < (3,2):
            _w("graphql.subscription.subscribe", "execute", _traced_operation("graphql.execute"))
        else:
            _w("graphql.execution.subscribe", "execute", _traced_operation("graphql.execute"))
      ``` 
- write tests for executing graphql subscriptions (graphql-core 2 and graphql-core 3)

## Checklist
- [ ] Added to the correct milestone.
- [ ] Tests provided or description of manual testing performed is included in the code or PR.
- [ ] Library documentation is updated.
- [ ] [Corp site](https://github.com/DataDog/documentation/) documentation is updated (link to the PR).

### CODE REVIEW

File: ddtrace/contrib/graphql/patch.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+from ddtrace.pin import Pin
+from ddtrace.vendor.wrapt import wrap_function_wrapper as _w
+
+from .. import trace_utils
+from ...ext import SpanTypes
+
+
+config._add("graphql", dict(_default_service="graphql"))
+
+graphql_version_str = getattr(graphql, "version", "0.0.0")
```
Comment (brettlangdon) [inline ddtrace/contrib/graphql/patch.py:21]: is `graphql.version_info` available back to 1.1? might make this easier for you?

File: ddtrace/contrib/graphql/patch.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+graphql_version = tuple([int(i) for i in graphql_version_str.split(".")])
+
+
+def patch():
+    if getattr(graphql, "_datadog_patch", False):
+        return
+    elif graphql_version < (1, 1):
```
Comment (brettlangdon) [inline ddtrace/contrib/graphql/patch.py:28]:
```suggestion
    if getattr(graphql, "_datadog_patch", False) or graphql_version < (1, 1):
```

File: ddtrace/contrib/graphql/patch.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+        return
+
+    setattr(graphql, "_datadog_patch", False)
+    _u(graphql, "graphql")
+    _u(graphql, "graphql_sync")
+
+
+async def _traced_graphql(func, instance, args, kwargs):
+    return await _traced_graphql_sync(func, instance, args, kwargs)
```
Comment (brettlangdon) [inline ddtrace/contrib/graphql/patch.py:50]: does this work? `_traced_graphql_sync` isn't an async function... does it return a task?

File: docs/integrations.rst [hunk 4]
```diff
@@ -189,6 +189,13 @@ gevent
 .. automodule:: ddtrace.contrib.gevent
 
 
+.. _graphql:
+
+gevent
```
Comment (brettlangdon) [inline docs/integrations.rst:194]:
```suggestion
graphql
```

File: ddtrace/contrib/graphql/patch.py (hunk 1 above)
Comment (mabdinur) [reply ddtrace/contrib/graphql/patch.py:21]: Yup it's available. I'll use that instead

File: ddtrace/contrib/graphql/patch.py (hunk 3 above)
Comment (mabdinur) [reply ddtrace/contrib/graphql/patch.py:43]: I believe it should work. I can refactor this if it's hard to follow. I added a test case which uses an async resolver.

`graphql.graphql(schema, source, ..., awaitable=False)` can return either an Awaitable or an ExecutionResult object : https://github.com/graphql-python/graphql-core/blob/main/src/graphql/graphql.py#L92-L95.  While `graphql.graphql_sync` always returns an ExecutionResult object. 

Both methods return the result of `graphql_impl()` which is a synchronous function: https://github.com/graphql-python/graphql-core/blob/main/src/graphql/graphql.py#L152. I could just instrument this method instead but it's not part of the public API so it might change/break in a minor version.

Comment (brettlangdon) [reply ddtrace/contrib/graphql/patch.py:43]: apologies for assuming it didn't work, it did look a little off to me, maybe it is just the naming (async function awaits a sync function) that threw me off

File: tests/contrib/graphql/test_graphql.py [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+@pytest.mark.asyncio
+async def test_graphql_async(test_spans, test_schema_async, test_source):
+    result = await graphql.graphql(test_schema_async, test_source)
+    assert result.data == {"hello": "friend"}
+
+    spans = test_spans.get_spans()
+    assert len(spans) == 1
+    assert spans[0].name == "graphql"
+    assert spans[0].resource == test_source
+    assert spans[0].service == "graphql"
```
Comment (mabdinur) [inline tests/contrib/graphql/test_graphql.py:87]: I plan to increase test coverage if this general approach looks good. I'll also add a sample django-graphene app with snapshot tests

File: tests/snapshots/test_graphql.test_graphql_error.json [hunk 6]
```diff
@@ -0,0 +1,23 @@
+[[
+  {
+    "name": "graphql_sync",
```
Comment (mabdinur) [inline tests/snapshots/test_graphql.test_graphql_error.json:3]: In Span.span.set_exc_info(), `Span.error = 1` is only set if a Span has a [traceback](https://github.com/DataDog/dd-trace-py/blob/v1.1.1/ddtrace/span.py#L426). 

`graphql.GraphqlError` does not generate a traceback for some errors. Ex: (Cannot query field)[https://github.com/graphql-python/graphql-core/blob/v3.2.1/src/graphql/validation/rules/fields_on_correct_type.py#L56]. 

Even though this Span does not have a traceback I would still expect it be marked as error Span. In cases like this should we generate a "fake" one line traceback or the current behavior acceptable?

File: tests/subprocesstest.py [hunk 7]
```diff
@@ -112,6 +112,7 @@ def _run_test_in_subprocess(self, result):
             stdout=subprocess.PIPE,
             stderr=subprocess.PIPE,
             env=sp_test_env,
+            shell=True,
```
Comment (mabdinur) [inline tests/subprocesstest.py:115]: `tests/contrib/graphql/test_graphql_patch.py` only passes with [subprocess.Popen(.., shell=True)](https://docs.python.org/3/library/subprocess.html#popen-constructor)

File: ddtrace/contrib/graphql/patch.py [hunk 8]
```diff
@@ -0,0 +1,121 @@
+from typing import Union
+
+import graphql
+from graphql.language.source import Source
+from promise import Promise
```
Comment (mabdinur) [inline ddtrace/contrib/graphql/patch.py:5]: Remove this dependency from graphql-core v3. Promise is only available in v2

File: ddtrace/contrib/graphql/patch.py [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+    pin = Pin.get_from(graphql)
+    if not pin or not pin.enabled():
+        return func(*args, **kwargs)
+
+    resource = _get_source_str(args, kwargs)
+
+    with pin.tracer.trace(
+        name="graphql_sync",
+        resource=resource,
+        service=trace_utils.ext_service(pin, config.graphql),
```
Comment (mabdinur) [inline ddtrace/contrib/graphql/patch.py:61]: `int_service`

File: releasenotes/notes/add-graphql-core-support-fd20dcbebfcaa8c7.yaml [hunk 10]
```diff
@@ -0,0 +1,5 @@
+---
+features:
+  - |
+    Add tracing support for ``graphql-core``. Version 2.0+ is fully
```
Comment (mabdinur) [inline releasenotes/notes/add-graphql-core-support-fd20dcbebfcaa8c7.yaml:4]: align with node graphql implementation details

File: tests/snapshots/tests.contrib.graphql.test_graphql.test_graphql_error.json [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+       "span_id": 2,
+       "parent_id": 1,
+       "type": "web",
+       "metrics": {
+         "_dd.measured": 1
+       },
+       "duration": 110000,
+       "start": 1652688948723380000
+     },
+     {
```
Comment (mabdinur) [inline tests/snapshots/tests.contrib.graphql.test_graphql.test_graphql_error.json:38]: I did not add error tags to this PR. This query failed during the validate step and returned before `graphql.execute()` was called. Is this acceptable?

**Comment by EtienneLamoureux** [comment]
I'm very excited for this change. Will instrumentation of all GraphQL APIs come free with using `ddtrace-run` or will we have to annotate the resolver in some fashion?

File: ddtrace/contrib/graphql/patch.py [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+        validate_module = "graphql.backend.core"
+
+    resolve_module = "graphql.execution.execute"
+    # ExecutionContext.resolve_field was renamed to execute_field in graphql-core 3.2
+    resolve_func = "ExecutionContext.execute_field"
+    if graphql_version < (3, 0):
+        resolve_module = "graphql.execution.executor"
+        resolve_func = "resolve_field"
+    elif graphql_version < (3, 2):
+        resolve_func = "ExecutionContext.resolve_field"
```
Comment (Kyle-Verhoog) [inline ddtrace/contrib/graphql/patch.py:56]: Can we only do a single version check and then duplicate the patch methods? This seems to be trying to be too clever and will be hard to change if one of the patching functions no longer works for a version.

Also it's hard to read the code and know immediately which functions are patched.

File: ddtrace/contrib/graphql/patch.py [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+
+    Pin().onto(graphql)
+
+
+def unpatch():
+    pass
+
+
+def _traced_graphql(func, instance, args, kwargs):
+    pin = Pin.get_from(graphql)
```
Comment (Kyle-Verhoog) [inline ddtrace/contrib/graphql/patch.py:72]: We have a helper in trace_utils for this: `with_traced_module`

File: tests/snapshots/tests.contrib.graphql.test_graphql.test_graphql_error.json (hunk 11 above)
Comment (Kyle-Verhoog) [reply tests/snapshots/tests.contrib.graphql.test_graphql.test_graphql_error.json:38]: We should be flagging the span as an error with all the error information we can to help the user debug the app.

File: tests/contrib/graphql/test_graphql_patch.py [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+
+
+class TestGraphqlPatch(PatchTestCase.Base):
+    __integration_name__ = "graphql"
+    __module_name__ = "graphql"
+    __patch_func__ = patch
+    __unpatch_func__ = None
+
+    def assert_module_patched(self, graphql):
+        self.assert_wrapped(graphql.graphql)
```
Comment (Kyle-Verhoog) [inline tests/contrib/graphql/test_graphql_patch.py:13]: Can we check all the different paths of the `execute` function?

File: tests/snapshots/tests.contrib.graphql.test_graphql.test_graphql_error_v2_promise.json [hunk 15]
```diff
@@ -0,0 +1,57 @@
+[[
+  {
+    "name": "graphql.query",
+    "service": "graphql",
+    "resource": "{ invalid_schema }",
+    "trace_id": 0,
+    "span_id": 1,
+    "parent_id": 0,
+    "type": "web",
+    "meta": {
```
Comment (mabdinur) [inline tests/snapshots/tests.contrib.graphql.test_graphql.test_graphql_error_v2_promise.json:10]: `graphql.graphql()` returns  promise. Although an error occurred in this span this is not known until the promise resolves. Unlike the `test_graphql_error.json` snapshot this error is only marked in the validate span.

**Comment by mabdinur** [comment]
> I'm very excited for this change. Will instrumentation of all GraphQL APIs come free with using `ddtrace-run` or will we have to annotate the resolvers in some fashion?

Hi @EtienneLamoureux, thanks for reaching out. The current plan is to instrument all resolvers using `ddtrace-run` but this approach could be noisy for some customers. If we decide not to auto instrument all resolvers in the initial release of this integration you could use `@tracer.wrap()` to instrument individual resolvers. 

Would you like to test out a POC of this integration and let us know what you think? We'd greatly appreciate any feedback.

**Comment by mabdinur** [comment]
> Hi @mabdinur,
> 
> Manually tagging my resolvers is perfectly acceptable for my use-case, no problem.
> 
> And I'd gladly test a POC of this feature! What's the best way to get a hold of a build with it in it?

Hey @EtienneLamoureux, sorry about the delay. Here's a link to our public slack channel: https://chat.datadoghq.com/. Feel free to join and we can continue our conversation there. Thanks again helping us test the graphql integration 😊

File: tests/snapshots/tests.contrib.graphql.test_graphql.test_graphql_error_v2_promise.json (hunk 15 above)
Comment (mabdinur) [reply tests/snapshots/tests.contrib.graphql.test_graphql.test_graphql_error_v2_promise.json]: This was fixed by patching `graphql.execute_graphql` instead of `graphql.graphql` in graphql-core-2

File: ddtrace/contrib/graphql/patch.py [hunk 16]
```diff
@@ -0,0 +1,189 @@
...
+    # Patch execute functions exposed in the public api
+    _w("graphql", "execute", _traced_operation("graphql.execute"))
+    _w("graphql.execution", "execute", _traced_operation("graphql.execute"))
+    if graphql_version < (3, 0):
+        _w("graphql.execution.executor", "execute", _traced_operation("graphql.execute"))
+    elif graphql_version >= (3, 1):
+        _w("graphql", "execute_sync", _traced_operation("graphql.execute"))
+        _w("graphql.execution", "execute_sync", _traced_operation("graphql.execute"))
+
+    # patch all resolvers
```
Comment (mabdinur) [inline ddtrace/contrib/graphql/patch.py:79]: Instead of monkey patching `resolve_field` I could instead pass a trace [middleware](https://docs.graphene-python.org/en/latest/execution/middleware/) to graphql.execute(). This middleware would wrap all resolvers in a span when resolve_field is called. Both approaches produce equivalent instrumentation (although using a trace middleware could provide more accurate span durations). I think the monkey patching approach is easier to follow.

ex: 
```
import graphql

class TracedResolverMiddleware(object):
    def resolve(self, next, root, info, **args):
        pin = Pin.get_from(graphql)
        if not pin or not pin.enabled():
            return next(root, info, **args)

        with pin.tracer.trace(
            name="graphql.resolve", 
            resource=info.field_name,
            service=trace_utils.int_service(pin, config.graphql),
            span_type=SpanTypes.WEB,
        ) as span:
            _init_span(span)
            return next(root, info, **args)

# wraps all resolvers in a span
result = graphql.execute(schema, source, middleware=[TracedResolverMiddleware])
```

**Comment by mabdinur** [comment]
The graphql integration will be introduced in the following PR: https://github.com/DataDog/dd-trace-py/pull/3878
